OOps/array_ops.c

```diff
@@ -34,18 +40,31 @@ int32_t array_init(CSOUND *csound, ARRAYINIT *p)
       csound->InitError(csound, "%s",
                         Str("Error: no sizes set for array initialization"));
 
+  const char *atype = (arrayDat && arrayDat->arrayType && arrayDat->arrayType->varTypeName)
+                        ? arrayDat->arrayType->varTypeName : NULL;
+  int isAudioArray = (atype && atype[0] == 'a');
+
   for (i = 0; i < inArgCount; i++) {
-    if (UNLIKELY(MYFLT2LRND(*p->isizes[i]) <= 0)) {
-      return
-        csound->InitError(csound, "%s",
-                          Str("Error: sizes must be > 0 for array initialization"));
+    if (UNLIKELY(p->iargs[i] == NULL)) {
+      return csound->InitError(csound, "%s",
+                               Str("Error: NULL size pointer for array initialization"));
+    }
+    int v = MYFLT2LRND(*p->iargs[i]);
+    if (UNLIKELY(v <= 0)) {
+      // Special-case for audio arrays: size 0 means use ksmps (first dimension only)
+      if (!(isAudioArray && i == 0)) {
+        return csound->InitError(csound, "%s",
+                                 Str("Error: sizes must be > 0 for array initialization"));
+      }
     }
   }
 
   arrayDat->dimensions = inArgCount;
   arrayDat->sizes = csound->Calloc(csound, sizeof(int32_t) * inArgCount);
   for (i = 0; i < inArgCount; i++) {
-    arrayDat->sizes[i] = MYFLT2LRND(*p->isizes[i]);
+    int v = MYFLT2LRND(*p->iargs[i]);
+    if (isAudioArray && i == 0 && v <= 0) v = CS_KSMPS;
```

Comment:
```suggestion
    int v = (isAudioArray && i == 0 && MYFLT2LRND(*p->iargs[i]) <= 0)
              ? CS_KSMPS
              : MYFLT2LRND(*p->iargs[i]);
```

Reply:
classic readability improvement. I commited this with the change of iargs being isizes now: e2498cdb493f901b7d1ba1686f15861d832bba76